jaxlib/cusparse.cc

```diff
@@ -854,6 +854,36 @@ py::bytes BuildGtsv2Descriptor(int m, int n, int ldb) {
   return PackDescriptor(Gtsv2Descriptor{m, n, ldb});
 }
 
+bool ComputeAsyncMallocSupported() {
+  int driver_version;
+  ThrowIfErrorStatus(cudaDriverGetVersion(&driver_version));
+  if (driver_version < 11020) {
+    return false;
+  }
+
+  // This is not documented, but we have found experimentally that for driver
+  // 460.56 cudaMallocAsync works on {P,V,A}100 (sm{6,7,8}0), but does not work
+  // on Quadro M4000 (sm52).
+  int device;
+  ThrowIfErrorStatus(cudaGetDevice(&device));
+  cudaDeviceProp prop;
+  ThrowIfErrorStatus(cudaGetDeviceProperties(&prop, device));
+  return prop.major > 5;
+}
+
+int cachedAsyncMallocSupported = -1;
+
+bool AsyncMallocSupported() {
+  if (cachedAsyncMallocSupported == -1) {
+    // This may get computed by multiple threads and overriden in paralllel,
+    // but we expect all threads to compute the same value (XLA requires
+    // replicated computations to be launched on equivalent devices) so it does
+    // not matter which threads result is written.
+    cachedAsyncMallocSupported = ComputeAsyncMallocSupported();
```

Comment:
Benign UB like this one might not be terrible, but it's still UB. But this would be perfectly thread safe:
```cpp
bool AsyncMallocSupported() {
  static bool cached = ComputeAsyncMallocSupported();
  return cached;
}
```

Reply:
Also, you're basing your decision on the compute capability which might vary between devices, so we might want to use e.g. a static array of booleans (32 is probably more than enough).

Reply:
Fair points, your solution was much simpler (I was not aware that this was thread safe). Luckily it is no longer needed since we can actually determine the buffer size ahead of time and get XLA to allocate it. Much simpler 😄